### ISSUE TITLE
tecNO_ENTRY when trying to burn the NFT from minted account even though the tfBurnable flag is set

### ISSUE THREAD

**Issue by chalith** [XRPLF/xrpl.js#1947]
When i'm trying to burn a NFT which I've minted but now owned by an another account, the **NFTokenBurn** transaction gives **tecNO_ENTRY** even though the **tfBurnable** flag **(1)** is set when the NFT is minted.

**Steps:**

> 1. Did a **NFTokenMint** with **Flags: 1** (for **tfBurnable**) from account **"A"** without specifying an **Issuer**
> ```
> {
>     TransactionType: 'NFTokenMint',
>     Account: <address of account A>
>     URI: <uri>,
>     TokenTaxon: 0,
>     TransferFee: 0,
>     Flags: 1
> }
> ```
> 3. Created a sell offer (**NFTokenCreateOffer**) for the **NFT** from account **"A"**
> 4. Accepted the sell offer (**NFTokenAcceptOffer**) from account **"B"** (So now **NFT** is owned by account **"B"**)
> 5. Tried to burn the **NFT** (**NFTokenBurn**) from account **"A"**
>```
> {
>     TransactionType: 'NFTokenBurn',
>     Account: <address of account A>,
>     TokenID: <the nft id>
> }
>```

Result:

> **NFTokenBurn** transaction gives **tecNO_ENTRY**

Expected Result:

> **tesSUCCESS**, since **tfBurnable** flag is set where the issuer (In our case the minted account) should be able to burn the **NFT**

**Comment by khancode** [comment]
@chalith I don't believe the issuer can burn the NFT when issuer no longer owns it; it doesn't sound like a valid use case too. @ledhed2222 do you know by chance?

**Comment by JST5000** [comment]
> @chalith I don't believe the issuer can burn the NFT when issuer no longer owns it; it doesn't sound like a valid use case too. @ledhed2222 do you know by chance?

Actually I believe allowing the issuer to burn a token even after it's been traded is what the **burnable** flag is supposed to allow. 

An example of when you'd want to do that is if you granted someone 24 access to a safety deposit box, and want to revoke that access afterwards. You could issue a burnable NFT, then burn the token after their time runs out. 

@chalith do you have a code snippet which reproduces the issue we can try? (That'll save us time experimenting with the problem)

**Comment by khancode** [comment]
> An example of when you'd want to do that is if you granted someone 24 access to a safety deposit box, and want to revoke that access afterwards. You could issue a burnable NFT, then burn the token after their time runs out.

Ah, that's valid use case. My mistake 🙂

**Comment by chalith** [comment]
> > @chalith I don't believe the issuer can burn the NFT when issuer no longer owns it; it doesn't sound like a valid use case too. @ledhed2222 do you know by chance?
> 
> Actually I believe allowing the issuer to burn a token even after it's been traded is what the **burnable** flag is supposed to allow.
> 
> An example of when you'd want to do that is if you granted someone 24 access to a safety deposit box, and want to revoke that access afterwards. You could issue a burnable NFT, then burn the token after their time runs out.
> 
> @chalith do you have a code snippet which reproduces the issue we can try? (That'll save us time experimenting with the problem)

Yes, here's my code snippet @JST5000 


```
const xrpl = require('xrpl');
const client = new xrpl.Client("wss://xls20-sandbox.rippletest.net:51233");
const A = {
    address: 'rPjirt4bi7F3gmyMhbfswYoebus88REZ1K',
    secret: 'ssUFW7Re7Ge9xBAMhJrdisi4LAok5'
}
const walletA = xrpl.Wallet.fromSeed(A.secret);

const B = {
    address: 'r4vE8ut2bnakgvWtp9mVxT7LEG52gVHnjP',
    secret: 'snhKXb8j7EKKHo847Sv15eYYn7kGP'
}
const walletB = xrpl.Wallet.fromSeed(B.secret);

async function burnTest() {
    await client.connect();

    const URI = '74657374757269'

    // Mint the NFT from account A.
    const mintRes = await client.submitAndWait({
        TransactionType: 'NFTokenMint',
        Account: A.address,
        URI: URI,
        TokenTaxon: 0,
        TransferFee: 0,
        Flags: 1
    }, { autofill: true, wallet: walletA });
    console.log('NFT mint from account A ended with', mintRes.result.meta.TransactionResult);

    const nfts = (await client.request({ command: 'account_nfts', account: A.address, ledger_index: "validated" }))?.result?.account_nfts;
    const nft = nfts?.find(n => n.URI === URI);
    const tokenId = nft.TokenID

    ////////////////////////////////////

    // Create offer for the NFT from account A.
    const offerRes = await client.submitAndWait({
        TransactionType: 'NFTokenCreateOffer',
        Account: A.address,
        TokenID: tokenId,
        Amount: "1",
        Flags: 1
    }, { autofill: true, wallet: walletA });
    console.log('NFT sell offer from account A ended with', offerRes.result.meta.TransactionResult);

    ////////////////////////////////////
    
    // Accept offer from account B.

    // Get sell offer created by account A //
    const offers = (await client.request({ command: 'account_objects', account: A.address }))?.result?.account_objects?.filter(o => o.LedgerEntryType === 'NFTokenOffer');
    const offer = offers?.find(o => o.Flags === 1 && o.Owner === A.address && o.TokenID === tokenId);

    const acceptRes = await client.submitAndWait({
        TransactionType: 'NFTokenAcceptOffer',
        Account: B.address,
        SellOffer: offer.index
    }, { autofill: true, wallet: walletB });
    console.log('NFT accept offer from account B ended with', acceptRes.result.meta.TransactionResult);

    ///////////////////////////////////

    // Burn the transfered NFT from account A.

    const burnRes = await client.submitAndWait({
        TransactionType: 'NFTokenBurn',
        Account: A.address,
        TokenID: tokenId
    }, { autofill: true, wallet: walletA });
    console.log('NFT burn offer from account A ended with', burnRes.result.meta.TransactionResult);

    await client.disconnect();
}

burnTest().catch(console.error)
```

**Comment by JST5000** [comment]
Hey @chalith, update on this issue :)

It seems that there's another field needed in NFTokenBurn when burning from the original account. You need to add the `Owner` field, which specifies which account currently owns the NFToken you want to burn. Otherwise `rippled` doesn't know where to look for the token, and defaults to checking the NFTokenPages of the `Account` field. (In this case, searching through A, and returning `tecNO_ENTRY` when it sees there is no token with that ID in A's NFTokenPages)

I just opened a PR to add that field to xrpl.js, and will be following up to do the same for the other client libraries + updating the xrpl.org documentation. 

Thanks for the clean repro code, it really helped!

**Comment by chalith** [comment]
> Hey @chalith, update on this issue :)
> 
> It seems that there's another field needed in NFTokenBurn when burning from the original account. You need to add the `Owner` field, which specifies which account currently owns the NFToken you want to burn. Otherwise `rippled` doesn't know where to look for the token, and defaults to checking the NFTokenPages of the `Account` field. (In this case, searching through A, and returning `tecNO_ENTRY` when it sees there is no token with that ID in A's NFTokenPages)
> 
> As a temporary workaround if you're using typescript, you can add that field to the `NFTokenBurn` object and just use a `// @ts-expect-error` comment above it and it'll work as expected.
> 
> I just opened a PR to add that field to xrpl.js, and will be following up to do the same for the other client libraries + updating the xrpl.org documentation.
> 
> Thanks for the clean repro code, it really helped!

Thank you @JST5000 for quickly attending to the issue. Appreciate your support on this :+1:  :)

**Comment by ckniffen** [comment]
Closed by #369
